### PR TITLE
Disable Self Sign-up

### LINKED ISSUE #2049: Make signup feature configurable
We should enable OpenFn deployment to enable/disable new user sign up on their platform. Currently this is set to happen by default and anyone can sign up from app.openfn.org. A fix to consider here will be making this configurable via the ENV so that when disabled, the signup text/link and route will be disabled on Lightning.

### PR DESCRIPTION
**Pull request by elias-ba**
## Validation Steps

To validate the changes made in your pull request, you can outline the following steps:

### Validation Steps

1. **Set Environment Variable:**
   - In your `.env` file, add or modify the following line:
     ```sh
     ALLOW_SIGNUP=false
     ```
   - Alternatively, modify `runtime.exs` to set the environment variable programmatically:
     ```elixir
     config :lightning,
       :disable_registration,
       env!("ALLOW_SIGNUP", &Utils.ensure_boolean/1, false)
     ```

2. **Restart the Application:**
   - Restart your Phoenix application to apply the configuration changes.

3. **Validate `/users/register` Route:**
   - Open your web browser and try to access the `/users/register` route.
   - **Expected Result:** The page should return a 404 error page, indicating that the route is inaccessible.

4. **Validate the Login Page:**
   - Open your web browser and navigate to the login page.
   - **Expected Result:** The "Register" button should be hidden from the login page, indicating that user registrations are disabled.

5. **Change `ALLOW_SIGNUP` to `true`:**
   - Set the `ALLOW_SIGNUP` value to `true` either in the `.env` file or `runtime.exs`.
   - Reload environment variables and restart the application.

6. **Re-validate `/users/register` Route:**
   - Access the `/users/register` route again.
   - **Expected Result:** The page should now be accessible, indicating that registrations are enabled.

7. **Re-validate the Login Page:**
   - Revisit the login page.
   - **Expected Result:** The "Register" button should now be visible.

## Notes for the reviewer

This PR introduces a plug that conditionally handles access to the /users/register route based on a configuration setting.

### Key Changes

- New Plug: Added `BlockRoutes` plug to conditionnaly block routes. The plug is generic enough to block any given route.
- Conditional 404 Response: Configured the plug to block the `/users/register` page if `:allow_signup` is false, the plug renders a 404 error page.
- Proceed when allowed: If `:allow_signup` is true, the plug allows the request to proceed as usual.

### Details

The plug checks takes a configuration which is a set of route, flag to check, and error message. It then checks if the incoming route matches the given route to block and if the flag is existent and truthy to return the given error message.

```
    plug LightningWeb.Plugs.BlockRoutes, [
      {
        "/users/register",
        :allow_signup,
        "Self-signup has been disabled for this instance. Please contact the administrator."
      }
    ]
```

### Testing

Added unit tests for the BlockRoutes plug to verify correct rendering of error message when routes are blocked, normal processing when they are allowed.

## Related issue

Fixes #2049 

## Review checklist

- [x] I have performed a **self-review** of my code
- [x] I have verified that all appropriate **authorization policies** (`:owner`, `:admin`, `:editor`, `:viewer`) have been implemented and tested
- [x] If needed, I have updated the **changelog**
- [ ] Product has **QA'd** this feature

### CODE REVIEW

**Review by taylordowns2000 — CHANGES_REQUESTED** [review]
The first two are stylistic changes. I'm happy to be overruled. The third one is a must.

1. Call the variable `ALLOW_SIGNUP` and make it default to `true`. Right now you've got a whole bunch of code that says something along the lines of `allow_signup = !Application.get_env('DISABLE_REG')`. That feels confusing both because it's different terminology (signup vs registration?) and because you're using that `!`.
2. Do we need a plug? I'd prefer that the page still exists but we conditionally render the signup form. When the form isn't there `if host == openfn.org, do: "Register at https://www.openfn.org/register", else: "Self-signup has been disabled for this instance. Contact your administrator for an account."`
3. Please add the new optional ENV to the [deployment](https://github.com/OpenFn/lightning/blob/main/DEPLOYMENT.md) docs.

File: test/lightning_web/plugs/register_gatekeeper_test.exs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  describe "call/2 when registration is enabled" do
+    setup do
+      Application.put_env(:lightning, :disable_registration, false)
+      :ok
+    end
+
+    test "passes through for /users/register" do
+      conn = conn(:get, "/users/register") |> RegisterGatekeeper.call([])
+
+      assert conn.status != 404
```
Comment (christad92) [inline test/lightning_web/plugs/register_gatekeeper_test.exs:41]: When a user tries to visit the sign up page, we should return a message saying "Sign up has been disabled for this instance, contact the administrator."

**Review by christad92 — COMMENTED** [review]
This is good. rather than returning a 404, we should tell the user that the feature has been disabled and they should contact the administrator. 

"Sign up has been disabled for this instance, please contact the administrator"

File: .env.example [hunk 2]
```diff
@@ -150,6 +150,9 @@
 # Give this variable the value of true if you want the system to create a sample project for a new registered user
 # INIT_PROJECT_FOR_NEW_USER=false
 
+# Give this variable the value of true if you want the system to prevent users from accessing the registration page
+# DISABLLE_REGISTRATION=false
```
[the remark this answers is not in the document]
Comment (elias-ba) [reply .env.example]: Done

File: test/lightning_web/plugs/register_gatekeeper_test.exs (hunk 1 above)
Comment (elias-ba) [reply test/lightning_web/plugs/register_gatekeeper_test.exs:41]: Done

**Review by christad92 — APPROVED** [review]
Looks good to me. Thanks

**Review by stuartc — APPROVED** [review]

**Review by stuartc — APPROVED** [review]

**Pull request merged**